### PR TITLE
add check to `@aws-amplify/backend` to throw early in a browser context

### PR DESCRIPTION
**Pull request by josefaidt**
## Problem

when folks attempt to use this in client-side applications they receive a cryptic error

```
Error: Can't resolve 'crypto' in '/Users/josef/my-amplify-app/node_modules/@aws-amplify/backend/lib'
```

**Issue number, if available:**

## Changes

This PR introduces an early check for browser globals and throws early if they're available.

I considered splitting this out into its own utility and tested outside the context of the secret utility, or lifting to the entrypoint of `@aws-amplify/backend` to prohibit use of any infra-related APIs client-side.

**Corresponding docs PR, if applicable:**

## Validation

added tests

## Checklist

- [x] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/backend/src/secret.ts [hunk 1]
```diff
@@ -27,6 +27,10 @@ import { BackendSecretFetcherFactory } from './engine/backend-secret/backend_sec
  *          ```
  */
 export const secret = (name: string): BackendSecret => {
+  // @ts-expect-error - this is a runtime check to prevent client-side usage
```
Comment (sobolk) [inline packages/backend/src/secret.ts:30]: could you try https://stackoverflow.com/questions/73261451/how-to-detect-if-code-runs-in-browser-in-typescript ?

File: packages/backend/src/secret.ts [hunk 2]
```diff
@@ -27,6 +27,10 @@ import { BackendSecretFetcherFactory } from './engine/backend-secret/backend_sec
...
+  if (typeof window !== 'undefined' || typeof document !== 'undefined') {
+    throw new Error('Secrets are not supported in client-side applications');
```
Comment (sobolk) [inline packages/backend/src/secret.ts:32]: Regarding `client-side` . The CLI, IaC is kind of "client side". Should we just say `browser` ?

Comment (josefaidt) [reply packages/backend/src/secret.ts]: that sounds fine, updated

File: packages/backend/src/is_browser.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+/**
+ * simple check to determine whether we are in a browser
```
Comment (sobolk) [inline packages/backend/src/is_browser.ts:2]:
```suggestion
 * Checks whether we are in a browser
```

**Review by sobolk — DISMISSED** [review]

**Review by sobolk — APPROVED** [review]

**Review by rtpascual — APPROVED** [review]

**Pull request merged**